### DISCUSSION TITLE
Add a transport that requests a proof of work to incoming connections if necessary

### DISCUSSION THREAD

**Discussion by tomaka** [libp2p/rust-libp2p#1972] [Ideas]
In order to prevent a node from being flooded with incoming connections, we can request a proof of work from incoming nodes. The difficulty of the proof would be proportional to the number of nodes that are currently attempting connection.

A timeout would of course be required.

The details of this scheme may be a bit tricky and should be figured out.

In details, once a certain threshold has been reached, the listener would cease to support all protocols except for `/libp2p/pow/1.0.0` (or a similar name). Once this protocol has been successfully negotiated and the PoW fulfilled, the dialer would request the actual protocol (eg. secio) on "top of it".

**Comment by CPerezz** [comment]
Is the goal to prevent DDOSing?
If so, there's an interesting paper on P2P networking [kadkast](https://eprint.iacr.org/2019/876.pdf) which uses PoW for the ID generation which is at the same time tied to the external IP that you have (so that you can't pre-compute the PoW). 

If it isn't, and it just wants to prevent overloading of the node, the first thing I would suggest for sure is to avoid any hashing algorithm that is GPU-highly optimizable since then it would become a mess to stablish a threshold to set as a minimum PoW required. I would suggest to use something like Poseidon or Pedersen hashes (which although they are optimizable, the difference it's not simply overwhelming as it happens with the SHA family). 

I would be interested in discussing more about that and helping if this is a feature that's really desired in the lib!

**Comment by mxinden** [comment]
Hi @CPerezz,

Thanks for offering your help. Given that the initial proposal is quite dated, I don't think this is a feature that is really desired today. This of course should not stop you from making a proposal in case you are in need for such mechanism.
